### PR TITLE
SIP - Voluntary Exit Proposal

### PR DESCRIPTION
**Pull request by MatheusFranco99**

### CODE REVIEW

**Review by GalRogozinski — CHANGES_REQUESTED** [review]
Maybe add a few words on how operators unregister the validators

File: sips/voluntary_exit.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+**New Runner**  
+A new runner for voluntary exits will be created.
+
+Upon exit message being parsed from EL, each node taking part in the specific cluster will broadcast SignedExitMessage including a partial signature for the voluntary exit.
+Upon a quorum of partial signatures, each node will reconstruct a valid signature and broadcast it to the beacon chain
+
+
+# Proposed solution
+
+_Validator voluntary exit_ will be a new type of duty. It'll be similar to validator registration with no consensus and no post-consensus steps (just pre-consensus).
```
[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply sips/voluntary_exit.md]: Changed

File: sips/voluntary_exit.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+
+**Specification**
+
+**Contract**  
+Adds a new function accessible only by the validator's owner, pushes a voluntray exit event. 
+Can only be called once.
+
+_Why a contract interaction?_ Enable contract based applications to do contract->contract zero coordination exits.
+Cost of such interaction is pretty minimal considering the value of a single validator.
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply sips/voluntary_exit.md:18]: My only thought at the moment that we require BLS2Execution address to be set when registering a validator via DKG.
Then the user behind the validator can use the withdrawal credentials to identify against our contract.

Or add a specialized credential that we use only with DKG?

**Review by GalRogozinski — COMMENTED** [review]
In order to handle the case where f+1 nodes were down or disconnected/eclipsed:
We can do the following. 

*Disconnected case*:
As long as 2F+1 aggregation of sigs wasn't achieved, refire the msg every epoch.

*Down case*:
Upon bootstrap use `getPastEvents` to find contract exit events. Check against the known active validators that are persisted by the node,  or query the beacon node to see whether the validators are in the exit queue.
If an exit event was found, but validator isn't in exit queue or exited refire message.


*Edit*:
Since the contract interaction is cheap it is not so bad if the validator will simply retrigger the exit. 
Since via UI/api the health of the operators can be checked, failures should be rare.

File: sips/voluntary_exit.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+**Networking**  
+A dedicated subnet for voluntary exits will be added to the network for all exit related communication between operators.
+All operators subscribe to the network.
```
[the remark this answers is not in the document]
Comment (alonmuroch) [reply sips/voluntary_exit.md:22]: idk, seems nicer to separate it no? Not a must

[the remark this answers is not in the document]
Comment (GalRogozinski) [reply sips/voluntary_exit.md:22]: @MatheusFranco99
Please remove the dedicated subnet

[the remark this answers is not in the document]
Comment (MatheusFranco99) [reply sips/voluntary_exit.md]: Done.

**Comment by GalRogozinski** [comment]
Do we know how to do MsgValidation on this @MatheusFranco99?

**Comment by MatheusFranco99** [comment]
In message validation, we are not verifying Voluntary exit messages yet. We can add new rules later to support this message type.

File: sips/voluntary_exit.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+**Summary**
+Describes how zero coordination voluntary exit works.  
+<em>Important to note that a user holding the validator’s private key can sign a voluntary exit as well.</em>
+
+
+**Specification**
+
+**Contract**  
+Adds a new function accessible only by the validator's owner, pushes a voluntray exit event. 
+Can only be called once.
```
Comment (lior-blox) [inline sips/voluntary_exit.md:15]: @alonmuroch Why only once?

Comment (alonmuroch) [reply sips/voluntary_exit.md:15]: Could be multiple times

File: sips/voluntary_exit.md [hunk 5]
```diff
@@ -53,4 +53,227 @@ type SignedExitMessage struct {
...
+```go
+// ValidatorVoluntaryExitCalls interface has all validator voluntary exit duty specific calls
+type ValidatorVoluntaryExitCalls interface {
+    // SubmitVoluntaryExit submits a validator voluntary exit
+    SubmitVoluntaryExit(voluntaryExit *phase0.SignedVoluntaryExit, sig phase0.BLSSignature) error
+}
+
+type BeaconNode interface {
+	// GetBeaconNetwork returns the beacon network the node is on
+	GetBeaconNetwork() types.BeaconNetwork
```
Comment (alonmuroch) [inline sips/voluntary_exit.md:69]: don't we have it already in the spec? why is this needeD?

Comment (MatheusFranco99) [reply sips/voluntary_exit.md:69]: Here, I mentioned the BeaconNode interface just to show how it updates.
But we don't have yet the specification of the VoluntaryExit call.

File: sips/voluntary_exit.md (hunk 4 above)
Comment (MatheusFranco99) [reply sips/voluntary_exit.md:15]: Updating to multiple times.

**Review by lior-blox — APPROVED** [review]

**Comment by lior-blox** [comment]
> In message validation, we are not verifying Voluntary exit messages yet. We can add new rules later to support this message type.

@MatheusFranco99 please add a TODO to msg validation

**Review by alonmuroch — APPROVED** [review]

**Review by lior-blox — APPROVED** [review]

File: sips/voluntary_exit.md (hunk 4 above)
Comment (mtabasco) [reply sips/voluntary_exit.md:15]: @MatheusFranco99 just confirming it could be multiple times. thanks.

Comment (MatheusFranco99) [reply sips/voluntary_exit.md:15]: Thanks @mtabasco
